README.MD
# SSD1306TUR
 **SSD1306TUR** is a library that allows you to write any **Turkish char** on **SSD1306** monochrome oled  displays freely. Based on Adafuit's SSD1306 oled Display driver can be found on [here](https://github.com/adafruit/Adafruit_SSD1306). Please be sure this lib is present in your system before use this lib.
 
 If you find any problem or bug on this library, please use [Issues](https://github.com/HakkanR/SSD1306TUR/issues) feature on github or contact me on my web page [devrelerim.com](https://www.devrelerim.com/2022/11/ssd1306tur-ssd1306-oled-ekran-turkce.htm).

 You can follow me on:
 * [Youtube](https://www.youtube.com/HakkanR)
 * [Twitter](https://www.twitter.com/HakkanR)
 * [Instagram](https://www.instagram.com/hakkan_r)
 * [FaceBook](https://www.facebook.com/HakkanR)
 
 ## Basic Usage
 include header file
 ```cpp
 #include <Wire.h>
 #include <SSD1306TUR.h>
 ```
 You should use fonts placed TrFonts directory in this repository by including like this:
```c
#include "TrFonts/FreeSansBold12pt7bTR.h"
```
define LCD specs fit your screen: 
```c
#define SCREEN_WIDTH 128
#define SCREEN_HEIGHT 64
#define SCREEN_ADDRESS 0x3C
```
initiate the object as any name you want, I choose display
```c
SSD1306TUR display(SCREEN_WIDTH, SCREEN_HEIGHT, &Wire);
```
begin screen
```c
display.begin(SSD1306_SWITCHCAPVCC, SCREEN_ADDRESS)
```
set text params and just write any Turkish letter by print, println or write functions.

```c
display.setTextColor(SSD1306_WHITE);
display.clearDisplay();
display.setTextSize(1);
display.setFont(&FreeSansBold12pt7bTR);
display.setCursor(20, 22);
display.print("Türkçe");
display.display(); 
```
## Dependencies 
* [Adafruit_SSD1306](https://github.com/adafruit/Adafruit_SSD1306)

## Download
You can download this on
* [Github](https://github.com/HakkanR/SSD1306TUR)
* Arduino Library Manager
* PlatformIO Libraries

## Available Fonts 
* FreeSansBold12pt7bTR.h
  
> more fonts will be added in the future, If you interested in creating new fonts to help this repository please feel free to contact me [hakkanr@gmail.com](mailto:hakkanr@gmail.com)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).



## [1.0.2] - 2023-06-13

### Added

- The default text color is set to SSD1306_WHITE to eliminate the need to use setTextColor function. 
- Corner Names for drawCircleHelper function.
- writeTr_128x64_i2c example ino file

## [1.0.1] - 2022-11-09

### Added

- Some additional files added for arduino library manager require

src/SSD1306TUR.h
/**
 * @file SSD1306TUR.h
 * @author Hakan ÖZMEN (hakkanr@gmail.com)
 * @brief
 * This is a library based on Adafruit's SSD1306 library for monochrome oled
 * displays, It targets to bring ability to use Turkish alphabet freely with these
 * oled screens. Adafruit_SSD1306 and Adafruit_GFX libraries are needed to use this
 * library but not enought to have Turkish letters, You should use font files in folder
 * TrFonts of this library. Notice that file names end with TR expression.
 *
 * @version 1.0.1
 * @date 2022-11-09
 * @section for more information please visit :
 * https://www.devrelerim.com/2022/11/ssd1306tur-ssd1306-oled-ekran-turkce.html
 * https://www.youtube.com/HakkanR
 * https://github.com/HakkanR/SSD1306TUR
 * @copyright
 * BDS License, all above text must be included any redistribution.
 *
 */
#ifndef SSD1306TUR_H_
#define SSD1306TUR_H_
#include <Adafruit_GFX.h>
#include <Adafruit_SSD1306.h>

// Corner Defines for Circle Helper Function
#define CORNLUP 1 // Top left corner of a circle
#define CORNLDN 8 // Buttom left corner of a circle
#define CORNRUP 2 // Top right corner of a circle
#define CORNRDN 4 // Buttom right corner of a circle

class SSD1306TUR : public Adafruit_SSD1306
{
public:
	// NEW CONSTRUCTORS -- recommended for new projects
	SSD1306TUR(uint8_t w, uint8_t h, TwoWire *twi,
			   int8_t rst_pin = -1, uint32_t clkDuring = 400000UL,
			   uint32_t clkAfter = 100000UL);
	SSD1306TUR(uint8_t w, uint8_t h, int8_t mosi_pin, int8_t sclk_pin,
			   int8_t dc_pin, int8_t rst_pin, int8_t cs_pin);
	SSD1306TUR(uint8_t w, uint8_t h, SPIClass *spi, int8_t dc_pin,
			   int8_t rst_pin, int8_t cs_pin, uint32_t bitrate = 8000000UL);

	// DEPRECATED CONSTRUCTORS - for back compatibility, avoid in new projects
	SSD1306TUR(int8_t mosi_pin, int8_t sclk_pin, int8_t dc_pin,
			   int8_t rst_pin, int8_t cs_pin);
	SSD1306TUR(int8_t dc_pin, int8_t rst_pin, int8_t cs_pin);
	SSD1306TUR(int8_t rst_pin = -1);
	virtual size_t write(uint8_t);
private:
	void resetFs();
private:
	uint8_t f195 = 0, f196 = 0, f197 = 0;
	uint8_t sent = 1;
};

#endif

src/SSD1306TUR.cpp
/**
 * @file SSD1306TUR.cpp
 * @author Hakan ÖZMEN (hakkanr@gmail.com)
 * @brief
 * * This is a library based on Adafruit's SSD1306 library for monochrome oled
 * displays, It targets to bring ability to use Turkish alphabet freely with these
 * oled screens. Adafruit_SSD1306 and Adafruit_GFX libraries are needed to use this
 * library but not enought to have Turkish letters, You should use font files in folder
 * TrFonts of this library. Notice that file names end with TR expression.
 * @section Dependencies :
 * This library depends on
 * https://github.com/adafruit/Adafruit-GFX-Library Adafruit_GFX, and
 * https://github.com/adafruit/Adafruit_SSD1306 Adafruit_SSD1306
 * be sure you have installed latest versions of these libs before
 * use this lib.
 * @version 1.0.1
 * @date 2022-11-09
 * @section for more information please visit :
 * https://www.devrelerim.com/2022/11/ssd1306tur-ssd1306-oled-ekran-turkce.html
 * https://www.youtube.com/HakkanR
 * https://github.com/HakkanR/SSD1306TUR
 * @copyright
 * BSD License, all above text must be included in any distribution.
 */
#include "SSD1306TUR.h"

SSD1306TUR::SSD1306TUR(uint8_t w, uint8_t h, TwoWire *twi = &Wire,
					   int8_t rst_pin, uint32_t clkDuring,
					   uint32_t clkAfter)
	: Adafruit_SSD1306(w, h, twi, rst_pin, clkDuring, clkAfter)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}
SSD1306TUR::SSD1306TUR(uint8_t w, uint8_t h, int8_t mosi_pin, int8_t sclk_pin,
					   int8_t dc_pin, int8_t rst_pin, int8_t cs_pin)
	: Adafruit_SSD1306(w, h, mosi_pin, sclk_pin, dc_pin, rst_pin, cs_pin)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}
SSD1306TUR::SSD1306TUR(uint8_t w, uint8_t h, SPIClass *spi, int8_t dc_pin,
					   int8_t rst_pin, int8_t cs_pin, uint32_t bitrate)
	: Adafruit_SSD1306(w, h, spi, dc_pin, rst_pin, cs_pin, bitrate)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}
SSD1306TUR::SSD1306TUR(int8_t mosi_pin, int8_t sclk_pin, int8_t dc_pin,
					   int8_t rst_pin, int8_t cs_pin)
	: Adafruit_SSD1306(mosi_pin, sclk_pin, dc_pin, rst_pin, cs_pin)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}
SSD1306TUR::SSD1306TUR(int8_t dc_pin, int8_t rst_pin, int8_t cs_pin)
	: Adafruit_SSD1306(dc_pin, rst_pin, cs_pin)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}

SSD1306TUR::SSD1306TUR(int8_t rst_pin)
	: Adafruit_SSD1306(rst_pin)
{
	textcolor = textbgcolor = SSD1306_WHITE;
}

size_t SSD1306TUR::write(uint8_t chr)
{
	if (chr == 195)
	{
		f195 = 1;
		sent = 0;
		return 1;
	}
	if (chr == 196)
	{
		f196 = 1;
		sent = 0;
		return 1;
	}
	if (chr == 197)
	{
		f197 = 1;
		sent = 0;
		return 1;
	}
	if (f196 == 1)
	{
		resetFs();
		if (chr == 177)
			chr = 0x7F;
		else if (chr == 176)
			chr = 0x80;
		else if (chr == 158)
			chr = 0x81;
		else if (chr == 159)
			chr = 0x82;
	}

	if (f195 == 1)
	{
		resetFs();
		if (chr == 182)
			chr = 0x83;
		else if (chr == 150)
			chr = 0x84;
		else if (chr == 156)
			chr = 0x85;
		else if (chr == 188)
			chr = 0x86;
		else if (chr == 135)
			chr = 0x87;
		else if (chr == 167)
			chr = 0x88;
	}
	if (f197 == 1)
	{
		resetFs();
		if (chr == 158)
			chr = 0x89;
		else if (chr == 159)
			chr = 0x8A;
	}
	return (sent ? Adafruit_SSD1306::write(chr) : 1);
}
void SSD1306TUR::resetFs()
{
	f195 = f196 = f197 = 0;
	sent = 1;
}

src/main.cpp
#include <Arduino.h>
#include "SSD1306TUR.h"
//#include "Fonts/FreeSerifItalic9pt7b.h"
//#include "Fonts/FreeSerif24pt7b.h"
#include "TrFonts/FreeSansBold12pt7bTR.h"

#define SCREEN_WIDTH 128 // OLED display width, in pixels
#define SCREEN_HEIGHT 32 // OLED display height, in pixels
#define OLED_RESET -1    // Reset pin # (or -1 if sharing Arduino reset pin)
#define SCREEN_ADDRESS 0x3C ///< See datasheet for Address; 0x3D for 128x64, 0x3C for 128x32
SSD1306TUR display(SCREEN_WIDTH, SCREEN_HEIGHT, &Wire, OLED_RESET);
void testscrolltext(void);
void setup()
{
  Serial.begin(115200);

  // SSD1306_SWITCHCAPVCC = generate display voltage from 3.3V internally
  if (!display.begin(SSD1306_SWITCHCAPVCC, SCREEN_ADDRESS))
  {
    Serial.println(F("SSD1306 allocation failed"));
    for (;;)
      ; // Don't proceed, loop forever
  }
  // Show initial display buffer contents on the screen --
  // the library initializes this with an Adafruit splash screen.
  //display.display();
  //delay(2000); // Pause for 2 seconds

  // Clear the buffer
  display.clearDisplay();
  testscrolltext();
}

void loop() {
  // put your main code here, to run repeatedly:
}
void testscrolltext(void)
{
  display.setTextColor(SSD1306_WHITE);
  display.clearDisplay();
  display.setTextSize(1);
  display.println("       SSD1306");
  display.setFont(&FreeSansBold12pt7bTR);
  display.setCursor(20, 22);
  display.print("Türkçe");
  display.display(); // Show initial text
  delay(3000);
  

  // Scroll in various directions, pausing in-between:
  // display.startscrollright(0x00, 0x0F);
  // delay(2000);
  // display.stopscroll();
  // delay(1000);
  // display.startscrollleft(0x00, 0x0F);
  // delay(2000);
  // display.stopscroll();
  // delay(1000);
  // display.startscrolldiagright(0x00, 0x07);
  // delay(2000);
  // display.startscrolldiagleft(0x00, 0x07);
  // delay(2000);
  // display.stopscroll();
  // delay(1000);
}